presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java

```diff
@@ -138,20 +138,15 @@ public synchronized void start()
 
     public synchronized void abort(TaskStatus taskStatus)
     {
-        updateTaskInfo(taskInfo.get().withTaskStatus(taskStatus));
+        updateTaskInfo(getTaskInfo().withTaskStatus(taskStatus));
         // For aborted tasks we do not care about the final task info, so stop the info fetcher
         stop();
     }
 
     public synchronized void taskStatusDone(TaskStatus taskStatus)
     {
-        if (running && !isDone(getTaskInfo())) {
-            // try to get the last task status from the remote task
-            taskStatusDone.set(taskStatus);
-        }
-        else {
-            getTaskInfo().withTaskStatus(taskStatus);
-        }
+        taskStatusDone.set(taskStatus);
+        updateTaskInfo(getTaskInfo().withTaskStatus(taskStatus));
```

Comment:
`taskStatusDone` indicates to the TaskInfoFetcher that the task is actually complete, but we still do not have the final stats and we need to make one more call to get the final stats. It was intentional to not call `updateTaskInfo` here. 

I am working on this (and surrounding) pieces of code to make it simpler and cleaner. I'll send out a PR soon.